civis/io/_tables.py

```diff
@@ -1094,3 +1168,122 @@ def split_schema_tablename(table):
+        cleaner_job = client.files.post_preprocess_csv(
+            file_id=fid,
+            in_place=False,
+            detect_table_columns=need_table_columns,
+            force_character_set_conversion=True,
+            hidden=hidden
+        )
+        cleaning_futures.append(run_job(cleaner_job.id, client=client))
+    return cleaning_futures
+
+
+def _process_cleaning_results(cleaning_futures, client, headers,
+                              need_table_columns, delimiter):
+    cleaned_file_ids = []
+
+    done, still_going = concurrent.futures.wait(
+        cleaning_futures, return_when=concurrent.futures.FIRST_COMPLETED
+    )
+
+    # Set values from first completed file cleaning - other files will be
+    # compared to this one. If inconsistencies are detected, raise an error.
+    first_completed = done.pop()
+    output_file = client.jobs.list_runs_outputs(
+            first_completed.job_id,
+            first_completed.run_id
+        )[0]
+    detected_info = client.files.get(output_file.object_id).detected_info
+    table_columns = (detected_info['tableColumns'] if need_table_columns
+                     else None)
+    if headers is None:
+        headers = detected_info['includeHeader']
+    elif headers != detected_info['includeHeader']:
+        raise CivisImportError('Mismatch between detected and provided '
+                               'headers - please ensure all imported files '
+                               'either have a header or do not.')
+    if delimiter is None:
+        delimiter = detected_info['columnDelimiter']
+    elif delimiter != detected_info['columnDelimiter']:
+        raise CivisImportError('Provided delimiter "{}" does not match '
+                               'detected delimiter for {}: "{}"'.format(
+                                   delimiter,
+                                   output_file.id,
+                                   detected_info["columnDelimiter"])
+                               )
+
+    compression = detected_info['compression']
+
+    cleaned_file_ids.append(output_file.object_id)
+
+    for result in concurrent.futures.as_completed(done | still_going):
+        output_file = client.jobs.list_runs_outputs(
+            result.job_id,
+            result.run_id
+        )[0]
+        detected_info = client.files.get(output_file.object_id).detected_info
+        if need_table_columns:
+            file_columns = detected_info['tableColumns']
+            if table_columns != file_columns:
+                raise CivisImportError('All files should have the same '
+                                       'schema. Expected {}'
+                                       'but file {} has {}'.format(
+                                           table_columns,
+                                           output_file.object_id,
+                                           file_columns)
+                                       )
+        if headers != detected_info['includeHeader']:
```

Comment:
This seems like a lot of repeated code from the checks on the first file. You could check for `None` separately in the first set of conditionals, then put this set of checks into a helper function that gets called here and above.

Reply:
John had suggested a refactor to this end in our discussion, but I pushed back then as the repetition didn't bother me as much. However, now that both you and he have raised it, I will move forward with this refactor.